libcudacxx/include/cuda/buffer

```diff
@@ -0,0 +1,171 @@
+//!    <libcudacxx-extended-api-memory-resources-resource>`. It is the users resposibility to ensure the lifetime of the
+//!    resource exceeds the lifetime of the buffer.
+//!
+//! @endrst
+template <class _Tp, class... _Properties>
+class uninitialized_buffer
+{
+private:
+  _CUDA_VMR::resource_ref<_Properties...> __mr_;
+  size_t __count_ = 0;
+  void* __buf_    = nullptr;
+
+  //! @brief Determines the allocation size given the alignment and size of @p _Tp
+  _CCCL_NODISCARD static constexpr size_t __get_allocation_size(const size_t __count) noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    return (__count * sizeof(_Tp) + (__alignment - 1)) & ~(__alignment - 1);
+  }
+
+  //! @brief Determines the properly aligned start of the buffer given the alignment and size of @p _Tp
+  _CCCL_NODISCARD constexpr _Tp* __get_data() const noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    size_t __space               = __get_allocation_size(__count_);
+    void* __ptr                  = __buf_;
+    return reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space));
+  }
+
+public:
+  //! @brief Constructs a ``cuda::uninitilaized_buffer`` from a memory resource and a number of elements
+  //! @param __mr The memory resource to allocate the buffer with.
+  //! @param __count The desired size of the buffer.
+  uninitialized_buffer(_CUDA_VMR::resource_ref<_Properties...> __mr, const size_t __count)
+      : __mr_(__mr)
+      , __count_(__count)
+      , __buf_(__mr_.allocate(__get_allocation_size(__count_)))
+  {}
+
+  uninitialized_buffer(const uninitialized_buffer&)            = delete;
+  uninitialized_buffer& operator=(const uninitialized_buffer&) = delete;
+  uninitialized_buffer(uninitialized_buffer&& __other) noexcept
+      : __mr_(__other.__mr_)
+      , __count_(__other.__count_)
+      , __buf_(__other.__buf_)
+  {
+    __other.__count_ = 0;
+    __other.__buf_   = nullptr;
+  }
+  uninitialized_buffer& operator=(uninitialized_buffer&& __other) noexcept
+  {
+    if (__buf_)
+    {
+      __mr_.deallocate(__buf_, __get_allocation_size(__count_));
+    }
+    __mr_            = __other.__mr_;
+    __count_         = __other.__count_;
+    __buf_           = __other.__buf_;
+    __other.__count_ = 0;
+    __other.__buf_   = nullptr;
+    return *this;
+  }
+
+  //! @brief Destroys a ``cuda::uninitilaized_buffer`` deallocating the buffer
+  ~uninitialized_buffer()
+  {
+    if (__buf_)
+    {
+      __mr_.deallocate(__buf_, __get_allocation_size(__count_));
+    }
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD constexpr _Tp* data() noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD constexpr const _Tp* data() const noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns the size of the buffer
```

Comment:
@bernhardmgruber: we should add `begin()` / `end()` and drop the span conversion
@ericniebler: seconds

Reply:
This is implemented